### PR TITLE
Added _repr_html_ for fonts

### PR DESCRIPTION
**Pull request by ambi7**
## PR Summary

Resolves [#18039](https://github.com/matplotlib/matplotlib/issues/18039)

Invoking in IPython
```python
import matplotlib.font_manager
from IPython.core.display import HTML

font = matplotlib.font_manager.fontManager.ttflist[0]       # any FontEntry

HTML(font._repr_html_())
```

## PR Checklist

**Tests and Styling**
- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [N/A] New features are documented, with examples if plot related.
- [N/A] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [N/A] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [N/A] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

### CODE REVIEW

**Review by anntzer — CHANGES_REQUESTED** [review]
As noted in #18039 I don't think this implementation is satisfactory; for example STIXGeneral is only provided by matplotlib on my computer and therefore trying to display in with this mechanism (which uses the browser's OS integration to lookup the font) silently renders a different font instead.  I *guess*(?) a strategy may be to use something like https://developer.mozilla.org/en-US/docs/Web/CSS/@font-face to explicitly specify a local font path instead.

**Comment by ambi7** [comment]
> As noted in #18039 I don't think this implementation is satisfactory; for example STIXGeneral is only provided by matplotlib on my computer and therefore trying to display in with this mechanism (which uses the browser's OS integration to lookup the font) silently renders a different font instead. I _guess_(?) a strategy may be to use something like https://developer.mozilla.org/en-US/docs/Web/CSS/@font-face to explicitly specify a local font path instead.

@anntzer Thank you for the review! I agree, the prior implementation will not render the correct font. I updated the PR to load the font using @font-face and updated tests. Can you take another look?

**Comment by anntzer** [comment]
Does this actually render the right font e.g. from a jupyter notebook?  From local testing it doesn't seem so?

**Comment by tacaswell** [comment]
I think it would be better to render to make a temporary `Figure` and put a base64 encodeding png (or a svg with the paths embeded) in the html repr (and also add `_repr_png_`).

**Comment by tacaswell** [comment]
@ambi7 Thank you for working on this!  I hope you are not discouraged by the initial feedback.

**Comment by ambi7** [comment]
> Does this actually render the right font e.g. from a jupyter notebook? From local testing it doesn't seem so?

Yes, font-face was having issues loading fonts from file when rendering from a jupyter notebook. I took @tacaswell suggestion and rendered the font via a png image. Looks like it is now rendering the right font by using a temporary `Figure`

**Comment by anntzer** [comment]
I like png as well here, but @timhoffm didn't (https://github.com/matplotlib/matplotlib/issues/18039#issuecomment-846591716).

**Comment by ambi7** [comment]
@timhoffm Any additional comments in regards to this implementation?

**Comment by timhoffm** [comment]
I can see the argument for images, that decouple the font rendering from the display.

So, this seems the defensive and sane approach. Still I'd like to see HiDPI-support for the rendered images, but that doesn't have to be necessarily part of this PR.

**Review by ambi7 — COMMENTED** [review]
resolved by creating _repr_png_ using Figure() to render custom font.

File: lib/matplotlib/font_manager.py [hunk 1]
```diff
@@ -380,6 +382,22 @@ def findSystemFonts(fontpaths=None, fontext='ttf'):
     return [fname for fname in fontfiles if os.path.exists(fname)]
 
 
+def fontentry_helper_repr_png_(fontent):
+    from matplotlib.figure import Figure  # Circular import.
+    fig = Figure()
+    font_path = Path(fontent.fname) if fontent.fname != '' else None
+    fig.text(0, 0, fontent.name, font=font_path)
+    with BytesIO() as buf:
+        fig.savefig(buf, bbox_inches='tight', transparent=True)
```
Comment (tacaswell) [inline lib/matplotlib/font_manager.py:391]: Is this enough to get high-dpi or does this just make things bigger?

```suggestion
        fig.savefig(buf, bbox_inches='tight', transparent=True, dpi=200)
```

**Review by tacaswell — APPROVED** [review]

**Comment by tacaswell** [comment]
Cleared @anntzer 's review because it was of the "ask the browser to render it" implementation.

File: lib/matplotlib/font_manager.py (hunk 1 above)
Comment (QuLogic) [reply lib/matplotlib/font_manager.py:391]: That will just make it bigger. The HTML needs to specify that the image is 2x, and I don't know how to tell that to Jupyter via the `_repr_png_` interface.

File: lib/matplotlib/tests/test_font_manager.py [hunk 2]
```diff
@@ -268,6 +270,27 @@ def test_fontcache_thread_safe():
                     f"{proc.returncode}.")
 
 
+def test_fontentry_dataclass():
+    fontent = FontEntry(name='font-name')
+
+    assert type(fontent.__doc__) == str
```
Comment (timhoffm) [inline lib/matplotlib/tests/test_font_manager.py:276]: Do we need to assert this?

File: lib/matplotlib/tests/test_font_manager.py [hunk 3]
```diff
@@ -268,6 +270,27 @@ def test_fontcache_thread_safe():
...
+
+    png = fontent._repr_png_()
+    html = fontent._repr_html_()
```
Comment (timhoffm) [inline lib/matplotlib/tests/test_font_manager.py:279]: Move to the respective places of usage. That's simpler to understand than defining stuff up front and only use it later.

File: lib/matplotlib/tests/test_font_manager.py (hunk 2 above)
Comment (ambi7) [reply lib/matplotlib/tests/test_font_manager.py]: No need, removed it

File: lib/matplotlib/tests/test_font_manager.py (hunk 3 above)
Comment (ambi7) [reply lib/matplotlib/tests/test_font_manager.py]: Done!

**Review by timhoffm — APPROVED** [review]

**Comment by timhoffm** [comment]
Thanks @ambi7!

**Pull request merged**